### PR TITLE
STM32 : protobuf + ethernet support

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Resolved Issues
#1008 and #1114 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/firmware_new/boards/frankie_v1/ethernetif.c
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/ethernetif.c:1]: ⚠️ this file is autogenerated

File: src/firmware_new/boards/frankie_v1/ethernetif.h
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/ethernetif.h:1]: ⚠️ this file is autogenerated

File: src/firmware_new/boards/frankie_v1/lwip.c
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/lwip.c:1]: ⚠️ this file is autogenerated

File: src/firmware_new/boards/frankie_v1/lwip.h
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/lwip.h:1]: ⚠️ this file is autogenerated

File: src/firmware_new/boards/frankie_v1/lwipopts.h
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/lwipopts.h:1]: ⚠️ this file is autogenerated

File: src/firmware_new/boards/frankie_v1/stm32h7xx_hal_msp.c
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/stm32h7xx_hal_msp.c:77]: ⚠️ this file is autogenerated

File: src/firmware_new/boards/frankie_v1/stm32h7xx_it.c
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/stm32h7xx_it.c:58]: ⚠️ this file is autogenerated

File: src/firmware_new/boards/frankie_v1/udp_echoserver.c
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/udp_echoserver.c:1]: ⚠️ this file is yoinked from LwIPs examples. It will be removed with the RTOS upgrade + proper API

**Review by akhilveeraghanta — COMMENTED** [review]
Just leaving some info about which files are autogenerated

**Comment by akhilveeraghanta** [comment]
Just wanted to start getting reviews on this stuff to wrap it up by the Jan 25th milestone (if the ublox chip continues to be easy and nothing comes up). This is pending full testing w/ the ublox, I have it working w/ a direct ethernet connection. This also does not have the multicast group stuff.

Ive left some pointers on which files are autogenerated, also recommend hiding whitespace

File: src/firmware_new/boards/frankie_v1/main.c
Comment (garethellis0) [inline src/firmware_new/boards/frankie_v1/main.c]: Why do we need this delay?

File: src/firmware_new/boards/frankie_v1/main.c
Comment (garethellis0) [inline src/firmware_new/boards/frankie_v1/main.c:126]: Just in case other people are using this project to do stuff, will this block and prevent other stuff from running? 

Also just so I can get a feel for where this is going, would this run in a RTOS task? At what frequency if so? Or would it be interrupt driven?

File: src/firmware_new/boards/frankie_v1/udp_echoserver.c
Comment (garethellis0) [inline src/firmware_new/boards/frankie_v1/udp_echoserver.c]: why do we need to "kill time" here? And how does this do so?

File: src/firmware_new/boards/frankie_v1/udp_echoserver.c
Comment (garethellis0) [inline src/firmware_new/boards/frankie_v1/udp_echoserver.c:123]: Just to be clear, this is the message giving the RX error on the computer side?

File: src/firmware_new/proto/control.proto
Comment (garethellis0) [inline src/firmware_new/proto/control.proto]: nit; how the robot responds to a message should not be part of the message. A message contains information, shouldn't dictate what to do with that information.

File: src/firmware_new/proto/BUILD
Comment (garethellis0) [inline src/firmware_new/proto/BUILD]: I don't want to block on this, but can you please file an issue for me and I can rewrite this to be a full-blown bazel rule (so it can use fields in the same sorta way we would with something like `cc_library`, etc.)

File: src/firmware_new/proto/nanopb.bzl
Comment (garethellis0) [inline src/firmware_new/proto/nanopb.bzl]: Does this break if the proto is not in the same folder as the one this macro is called from? Should we use absolute paths instead?

File: src/firmware_new/tools/communication/transfer_media/network_medium.cpp
Comment (garethellis0) [inline src/firmware_new/tools/communication/transfer_media/network_medium.cpp:37]: I think you want `std::rethrow_exception` (http://www.cplusplus.com/reference/exception/rethrow_exception/) here?

File: src/firmware_new/tools/communication/transfer_media/transfer_medium.h
Comment (garethellis0) [inline src/firmware_new/tools/communication/transfer_media/transfer_medium.h]: if you're just trying to enforce that these functions be overriden, should be able to just add `= 0` after them and then it should be compile-time enforced I think?

File: src/firmware_new/tools/send_proto_over_udp.cpp
Comment (garethellis0) [inline src/firmware_new/tools/send_proto_over_udp.cpp:38]: This architecture is pretty clean, nice job!

File: src/software/ai/hl/stp/tactic/tactic_visitor.h
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/tactic_visitor.h]: Did this get accidentally added? Doesn't seem relevant.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good! Don't see any major issues.

File: src/firmware_new/boards/frankie_v1/udp_echoserver.c
Comment (MathewMacDougall) [reply src/firmware_new/boards/frankie_v1/udp_echoserver.c:123]: As a followup, why is there an ACK at all? This seems more like an ACK for testing purposes rather than a protocol ACK (like TCP)?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looking good. Nothing major to add beyond Gareth's comments

File: src/firmware_new/boards/frankie_v1/BUILD
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/BUILD]: yeah, removed

File: src/firmware_new/boards/frankie_v1/.mxproject
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/.mxproject]: nope, nuked

File: src/firmware_new/boards/frankie_v1/main.c
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/main.c]: nuked it, was from earlier times when I was still debugging ethernet

File: src/firmware_new/boards/frankie_v1/main.c
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/main.c:126]: nope it returns after its processed one frame IIRC, it defn doesn't block forever. 

just checked some RTOS examples, it seems like they are using netconn, and it looks like its just a thread w/ an infinite loop

File: src/firmware_new/boards/frankie_v1/udp_echoserver.c
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/udp_echoserver.c]: im not even sure what that comment is anymore lol, nuked

File: src/firmware_new/boards/frankie_v1/udp_echoserver.c
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/udp_echoserver.c:123]: its just to get a sense of the rate we are receiving/decoding/encoding/sending. This whole file will be nuked after RTOS and this won't be a part of the API

File: src/firmware_new/proto/control.proto
Comment (akhilveeraghanta) [reply src/firmware_new/proto/control.proto]: removed

File: src/firmware_new/proto/BUILD
Comment (akhilveeraghanta) [reply src/firmware_new/proto/BUILD]: yeah it lacks Bazel "context" (TODO edit this w/ ticket number)

edit: #1133

File: src/firmware_new/proto/nanopb.bzl
Comment (akhilveeraghanta) [reply src/firmware_new/proto/nanopb.bzl]: yeah it does break but Bazel expects you to put a BUILD file in the folder w/ the src anyway so I figured it would be easier to be relative. it might be easier to enforce that rule than have to edit this path for every proto folder

File: src/firmware_new/tools/communication/transfer_media/network_medium.cpp
Comment (akhilveeraghanta) [reply src/firmware_new/tools/communication/transfer_media/network_medium.cpp:37]: https://stackoverflow.com/questions/19984686/boost-rethrow-exception-throws-exception-of-different-type

looks like boost doesn't let you do this by design, `throw` does rethrow the same exception

File: src/firmware_new/tools/communication/transfer_media/transfer_medium.h
Comment (akhilveeraghanta) [reply src/firmware_new/tools/communication/transfer_media/transfer_medium.h]: I tried this initially, but I wanted to be able to have a generic `TransferMedium` as a private member inside the communicator class, wasn't letting me do that when the functions where pure virtual

File: src/software/ai/hl/stp/tactic/tactic_visitor.h
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/tactic/tactic_visitor.h]: nuked, bad merge

File: src/firmware_new/boards/frankie_v1/BUILD
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/BUILD]: fixed

File: src/firmware_new/boards/frankie_v1/.mxproject
Comment (garethellis0) [reply src/firmware_new/boards/frankie_v1/.mxproject]: Could we add this to the `.gitignore` while we're here?

File: src/firmware_new/proto/nanopb.bzl
Comment (garethellis0) [reply src/firmware_new/proto/nanopb.bzl]: Ok, please just make a note of this in the the ticket for the full-blown rule from the above comment.

File: src/firmware_new/boards/frankie_v1/.mxproject
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/.mxproject]: added

File: src/firmware_new/tools/communication/transfer_media/transfer_medium.h
Comment (garethellis0) [reply src/firmware_new/tools/communication/transfer_media/transfer_medium.h]: Yeah, generally the approach here is to store the `TransferMedium` as a pointer (preferable a `std::unique_ptr`). Could we do that here to be consistent with the rest of the repo? That also gives us the bonus of compile-time guarantees.

File: src/firmware_new/boards/frankie_v1/.mxproject
Comment (garethellis0) [reply src/firmware_new/boards/frankie_v1/.mxproject]: Ah I see you've just done this. Thanks!

File: src/firmware_new/proto/nanopb.bzl
Comment (akhilveeraghanta) [reply src/firmware_new/proto/nanopb.bzl]: added

File: src/firmware_new/tools/communication/transfer_media/transfer_medium.h
Comment (akhilveeraghanta) [reply src/firmware_new/tools/communication/transfer_media/transfer_medium.h]: I switched over to `unique_ptrs` earlier, but didn't realize this was possible

switched to pure virtual, works well

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Review by MathewMacDougall — DISMISSED** [review]
nice, LGTM. Thanks for marking which files were autogenerated :pray:

File: src/firmware_new/boards/frankie_v1/BUILD
Comment (jonathanlew) [inline src/firmware_new/boards/frankie_v1/BUILD]: Can we glob these files at all?

File: src/firmware_new/proto/control.proto
Comment (jonathanlew) [inline src/firmware_new/proto/control.proto:3]: Is there an issue to properly define these messages?

File: src/firmware_new/tools/send_proto_over_udp.cpp
Comment (jonathanlew) [inline src/firmware_new/tools/send_proto_over_udp.cpp:39]: Should these numbers be constants or dynamic params?

**Review by jonathanlew — COMMENTED** [review]
Looks good! Just some minor comments

File: src/firmware_new/boards/frankie_v1/BUILD
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/BUILD]: we could but there are files that we don't need as well in that folder, and the rest of the BUILD file was listing them out so I just added to it

File: src/firmware_new/proto/control.proto
Comment (akhilveeraghanta) [reply src/firmware_new/proto/control.proto:3]: no there isn't, I think that will just come w/ whoever is going to use the `RobotCommunicator` class, the control msg is just a test msg like it says

File: src/firmware_new/tools/send_proto_over_udp.cpp
Comment (akhilveeraghanta) [reply src/firmware_new/tools/send_proto_over_udp.cpp:39]: it'll be up to where this is constructed, here its just for testing, eventually it will be a multicast group IP in which case it makes sense as a dynamic parameter w/ 4 multicast group ip options (for example) but for now and in this place its just raw until that's sorted out

File: src/firmware_new/proto/control.proto
Comment (jonathanlew) [reply src/firmware_new/proto/control.proto:3]: Fine

**Review by jonathanlew — DISMISSED** [review]

**Comment by akhilveeraghanta** [comment]
not sure why formatting check is failing, I get no diff when I run it locally

**Comment by garethellis0** [comment]
This should be a non-issue once https://github.com/UBC-Thunderbots/Software/pull/1132 is merged......

**Comment by akhilveeraghanta** [comment]
@jonathanlew @garethellis0 @MathewMacDougall need a re-review had to make some changes to the `nanopb.bzl` rule and ran fix-formatting hopefully everything passes now

**Comment by garethellis0** [comment]
Looks like some funky protobuf errors in the CI run?

**Review by henryymliu — COMMENTED** [review]
Nothing else to add. I'm curious about the host-side RX error; is it something like a malformed packet?

**Comment by akhilveeraghanta** [comment]
whoops that was a while back, forgot to update the desc, it doesn't have that issue anymore

**Comment by akhilveeraghanta** [comment]
Can one of you guys try this branch and see if you get the same build error, not sure what's causing this

**Comment by garethellis0** [comment]
So I can't repro this locally either, but looking at the `README` for `NanoPb` I see:
```
However, if you are using a git checkout or a plain source distribution, you need to provide your own version of protoc and the Google's protobuf library. On Linux, the necessary packages are protobuf-compiler and python-protobuf.
```

So I think we just have to install these ourselves.

Here's the diff that should install all the required deps:
```
diff --git a/environment_setup/setup_software.sh b/environment_setup/setup_software.sh
index 247c507c..f77af4e2 100755
--- a/environment_setup/setup_software.sh
+++ b/environment_setup/setup_software.sh
@@ -48,6 +48,12 @@ host_software_packages=(
     python3-yaml # yaml for cfg generation (Dynamic Parameters)
     python-minimal # This is required for bazel, we've seen some issues where
                    # the bazel install hasn't installed it properly
+    python-protobuf # This is required for the "NanoPb" library, which does not
+                    # properly manage this as a bazel dependency, so we have 
+                    # to manually install it ourselves
+    protobuf-compiler # This is required for the "NanoPb" library, which does not
+                      # properly manage this as a bazel dependency, so we have 
+                      # to manually install it ourselves
 )
 sudo apt-get install ${host_software_packages[@]} -y
 ```

It's an interesting idea actually, could reduce compile times if we didn't build protobuf ourselves for our code that depends on it. Super low priority though.

**Comment by akhilveeraghanta** [comment]
looks CI has those packages now, but it still seems to fail, 

Maybe worth building proto ourselves? Should resolve this, they probably have bazel BUILD files

**Comment by garethellis0** [comment]
Ok, I'm guessing that `python-protobuf` installs it for `python2`, but bazel uses `python3`. Try changing that patch to `python3-protobuf` and that might work.

**Comment by akhilveeraghanta** [comment]
@garethellis0 that did it, @MathewMacDougall @jonathanlew @garethellis0 need a re-approval

**Review by garethellis0 — APPROVED** [review]

**Review by MathewMacDougall — APPROVED** [review]
:fire: :signal_strength:

**Pull request merged**
